internal/pkg/storage/image.go

```diff
@@ -489,14 +489,21 @@ func (svc *imageService) ResolveNames(systemContext *types.SystemContext, imageN
 		}
 	}
 	// This to prevent any image ID to go through this routine
-	_, err := reference.ParseNormalizedNamed(imageName)
+	ref, err := reference.ParseNormalizedNamed(imageName)
 	if err != nil {
 		if strings.Contains(err.Error(), "cannot specify 64-byte hexadecimal strings") {
 			return nil, ErrCannotParseImageID
 		}
 		return nil, err
 	}
 
+	// Use digest instead of tag when both provided (for docker compatibility)
+	_, isTagged := ref.(reference.NamedTagged)
+	canonical, isDigested := ref.(reference.Canonical)
+	if isTagged && isDigested {
+		imageName = canonical.Name() + "@" + canonical.Digest().String()
+	}
+
```

Comment:
I have the feeling that this logic should somehow go into containers/image, also considering that we want to move to the libpod image backend in the near future. (not blocking)

Reply:
See e.g. the discussion in https://github.com/cri-o/cri-o/issues/2351 (and notably https://github.com/cri-o/cri-o/issues/2351#issuecomment-492726855 , which does suggest this could end up in Libpod).

One part of this is the ambiguity of intent/guarantees, another is that there are, sadly, quite a few manually-written incorrect parsers (“Is there a `@`? If so, the digest is everything after the first `:`”) that keep being added throughout the ecosystem. Rejecting :tag@digest in c/image somewhat limits the damage of such incorrect parsers. (OTOH, arguably, allowing :tag@digest could motivate maintainers of the code that use such parsers to fix the code.)

Reply:
@saschagrunert previously I tried to solve it in [containers/image#784](https://github.com/containers/image/pull/784). However I realized what @mtrmac pointed out and that it would make the library unnecessarily more complicated or inconsistent while actually being a non-solution just for docker-compatibility.